Microsoft.O365.Security.Native.ETW/EventRecordMetadata.hpp

```diff
@@ -153,6 +153,68 @@ namespace Microsoft { namespace O365 { namespace Security { namespace ETW {
             return dest;
         }
 
+#pragma endregion
+
+#pragma region ExtendedData
+
+        /// <summary>
+        /// If the event's extended data contains an Argon container ID, retrieve it.
+        /// Can be expensive, avoid calling more than once per event.
+        /// </summary>
+        /// <returns>
+        /// A Guid representing the Argon container ID if the data is present. Null it's not present. 
+        /// Throws a ContainerIdFormatException if the container ID is present but parsing fails.
+        /// </returns>
+        virtual System::Nullable<System::Guid> GetContainerId()
+        {
+            // We are expecting format "00000000-0000-0000-0000-0000000000000", 32 hex digits with 4 hyphens
+            constexpr size_t CONTAINER_ID_DATA_LENGTH_IN_BYTES = 36;
+
+            size_t extended_data_count = record_->ExtendedDataCount;
+            for (size_t i = 0; i < extended_data_count; i++)
+            {
+                EVENT_HEADER_EXTENDED_DATA_ITEM& extended_data = record_->ExtendedData[i];
+
+                if (extended_data.ExtType == EVENT_HEADER_EXT_TYPE_CONTAINER_ID)
+                {
+                    // Convert the non-null terminated, no-braces ASCII GUID into a null terminated, curly braces, wide string 
+                    // for parsing.
+                    std::wstring guid_string_buffer (L"{########-####-####-####-############}");
+                    assert(extended_data.DataSize == CONTAINER_ID_DATA_LENGTH_IN_BYTES);
+                    assert(guid_string_buffer.length() == 1 + CONTAINER_ID_DATA_LENGTH_IN_BYTES + 1);
+
+                    for (size_t c = 0; c < CONTAINER_ID_DATA_LENGTH_IN_BYTES; c++)
+                    {
+                        // Offset by 1 to account for curly brace
```

Comment:
is it mandatory that we return a GUID with curly braces? that's the `B` format for Guid.ToString (ref https://docs.microsoft.com/en-us/dotnet/api/system.guid.tostring?view=netframework-4.8) but not the default Guid format. seems like it's making this code more complex.

Reply:
I originally did not include curly braces, but from my testing it looked like CLSIDFromString fails with error CO_E_CLASSSTRING if those curly braces are not present.

Reply:
Is it mandatory to return a GUID type here? I wonder whether we should just return the string and avoid doing the extra work to validate the format.

Alternatively -- since you've already `assert`ed that `extended_data.DataSize` is the correct length -- you could simply call ConvertGuid on `extended_data.DataPtr`. It would be fine to add an overload of `ConvertGuid` for this. This code would be faster and simpler as a result.

Reply:
The underlying issue is that the ExtendedData buffer is a non-null-terminated string, which doesn't leave any real options for parsing or marshalling the data unless we append a null terminator first.

As we discussed separately, in order to avoid the performance impact of having to clone the string to a new buffer, I'll implement something to parse the GUID directly out of the buffer.

Reply:
Ok done - the latest commit implements a parser to read straight from the ExtendedData buffer and write into a GUID object.